sys/net/credman/credman.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_credman
+ * @{
+ *
+ * @file
+ * @brief       (D)TLS Credentials management module implementation
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ */
+
+#include "net/credman.h"
+#include "mutex.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static mutex_t _mutex = MUTEX_INIT;
+
+static credman_credential_t credentials[CREDMAN_MAX_CREDENTIALS];
+static unsigned used = 0;
+
+static int _find_credential_pos(credman_tag_t tag, credman_type_t type);
+static int _find_next_free_pos(void);
+
+int credman_add(const credman_credential_t *credential)
+{
+    assert(credential);
+    mutex_lock(&_mutex);
+    int pos = -1;
+    int ret = CREDMAN_ERROR;
+
+    if ((credential->type == CREDMAN_TYPE_EMPTY) ||
+        (credential->tag == CREDMAN_TAG_EMPTY)) {
+        DEBUG("credman: invalid credential type/tag\n");
+        ret = CREDMAN_INVALID;
+        goto end;
+    }
+    switch (credential->type) {
+    case CREDMAN_TYPE_PSK:
+        if ((credential->params.psk == NULL) ||
+            (credential->params.psk->key.s == NULL) ||
+            (credential->params.psk->key.len == 0)) {
+            DEBUG("credman: invalid PSK parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    case CREDMAN_TYPE_ECDSA:
+        if ((credential->params.ecdsa == NULL) ||
+            (credential->params.ecdsa->private_key == NULL) ||
+            (credential->params.ecdsa->public_key.x == NULL) ||
+            (credential->params.ecdsa->public_key.y == NULL)) {
+            DEBUG("credman: invalid ECDSA parameters\n");
+            ret = CREDMAN_INVALID;
+            goto end;
+        }
+        break;
+    default:
+        ret = CREDMAN_TYPE_UNKNOWN;
+        goto end;
+    }
+    pos = _find_credential_pos(credential->tag, credential->type);
+    if (pos >= 0) {
+        DEBUG("credman: credential with tag %d and type %d already exist\n",
+              credential->tag, credential->type);
+        ret = CREDMAN_EXIST;
+        goto end;
+    }
+    pos = _find_next_free_pos();
+    if (pos < 0) {
+        DEBUG("credman: no space for new credential\n");
+        ret = CREDMAN_NO_SPACE;
+        goto end;
+    }
+
+    credentials[pos] = *credential;
```

Comment:
In the description you say 

> It just holds the pointers to the credentials given by the user.

but here the value is being copied. So that the user must keep is the "*_params_t" structure _and_ the buffers pointed to by that structure.

Reply:
Thanks. I missed that. `credman_credential_t` initially have `*_params_t` itself in it, not just the pointer to it. I'll need to change that back.